File: PythonExample/NumPy/ex1/normalEqn.py
import numpy as np

'''
function [theta] = normalEqn(X, y)
%NORMALEQN Computes the closed-form solution to linear regression 
%   NORMALEQN(X,y) computes the closed-form solution to linear 
%   regression using the normal equations.

theta = zeros(size(X, 2), 1);

% ====================== YOUR CODE HERE ======================
% Instructions: Complete the code to compute the closed form solution
%               to linear regression and put the result in theta.
%

% ---------------------- Sample Solution ----------------------


theta = pinv(X' * X) * X' * y;

% -------------------------------------------------------------


% ============================================================

end
'''


# noinspection PyPep8Naming
def normal_eqn(X, y):
    temp = np.dot(X.T, X)
    temp = np.linalg.pinv(temp)
    theta = np.dot(np.dot(temp, X.T), y)

    return theta

File: ex1_linear_regression/normalEqn.m
function [theta] = normalEqn(X, y)
%NORMALEQN Вычисляет замкнутую форму решения для линейной регрессии
% NORMALEQN (X, y) вычисляет решение в замкнутой форме для линейного
% регрессии с использованием нормальных уравнений.

theta = zeros(size(X, 2), 1);

% ====================== ВАШ КОД ЗДЕСЬ ======================
% Инструкции: завершите код для вычисления решения в закрытой форме
% к линейной регрессии и положить результат в тета.
%

% ---------------------- Sample Solution ----------------------


theta = pinv(X' * X) * X' * y;
% theta = (X' * X)^-1 * X' * y;
% note: can also be written with pinv function: pinv(X' * X) * X' * y
% -------------------------------------------------------------


% ============================================================

end
